webapp/graphite/storage.py

```diff
@@ -80,60 +84,60 @@ def find(self, pattern, startTime=None, endTime=None, local=False):
-          def distance_to_requested_interval(node):
-            latest = sorted(node.intervals, key=lambda i: i.end)[-1]
-            distance = query.interval.start - latest.end
-            return distance if distance >= 0 else float('inf')
-
-          best_candidate = min(leaf_nodes, key=distance_to_requested_interval)
-          if distance_to_requested_interval(best_candidate) <= settings.FIND_TOLERANCE:
-            minimal_node_set.add(best_candidate)
+      # Calculate best minimal node set
+      minimal_node_set = set()
+      covered_intervals = IntervalSet([])
+
+      # If the query doesn't fall entirely within the FIND_TOLERANCE window
+      # we disregard the window. This prevents unnecessary remote fetches
+      # caused when carbon's cache skews node.intervals, giving the appearance
+      # remote systems have data we don't have locally, which we probably do.
+      now = int( time.time() )
+      tolerance_window = now - settings.FIND_TOLERANCE
+      disregard_tolerance_window = query.interval.start < tolerance_window
+      prior_to_window = Interval( float('-inf'), tolerance_window )
+
+      def measure_of_added_coverage(node, drop_window=disregard_tolerance_window):
+        relevant_intervals = node.intervals.intersect_interval(query.interval)
+        if drop_window:
+          relevant_intervals = relevant_intervals.intersect_interval(prior_to_window)
+        return covered_intervals.union(relevant_intervals).size - covered_intervals.size
+
+      nodes_remaining = list(leaf_nodes)
+
+      # Prefer local nodes first (and do *not* drop the tolerance window)
+      for node in leaf_nodes:
+        if node.local and measure_of_added_coverage(node, False) > 0:
+          nodes_remaining.remove(node)
+          minimal_node_set.add(node)
+          covered_intervals = covered_intervals.union(node.intervals)
+
+      while nodes_remaining:
+        node_coverages = [ (measure_of_added_coverage(n), n) for n in nodes_remaining ]
+        best_coverage, best_node = max(node_coverages)
+
+        if best_coverage == 0:
+          break
+
+        nodes_remaining.remove(best_node)
+        minimal_node_set.add(best_node)
+        covered_intervals = covered_intervals.union(best_node.intervals)
+
+      # Sometimes the requested interval falls within the caching window.
+      # We include the most likely node if the gap is within tolerance.
+      if not minimal_node_set:
+        def distance_to_requested_interval(node):
+          latest = sorted(node.intervals, key=lambda i: i.end)[-1]
+          distance = query.interval.start - latest.end
+          return distance if distance >= 0 else float('inf')
+
+        best_candidate = min(leaf_nodes, key=distance_to_requested_interval)
+        if distance_to_requested_interval(best_candidate) <= settings.FIND_TOLERANCE:
+          minimal_node_set.add(best_candidate)
+
+      if settings.REMOTE_STORE_MERGE_RESULTS and not local and remote_nodes:
```

Comment:
I should point out you're doing this inside the `for path, nodes` loop started [here](https://github.com/graphite-project/graphite-web/pull/1674/files#diff-57c0c4ffa18833a2e2ea55aab4309ffcR73) but these `remote_nodes` don't look to be filtered by `path`.

There's something I want to check up on...  If we're assuming [this branching](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-57c0c4ffa18833a2e2ea55aab4309ffcR83) is the cause of slow down, then we're also assuming there are overlaps in coverage with purely local nodes and that [this](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-57c0c4ffa18833a2e2ea55aab4309ffcR110) was working to reduce that overlaps.

However, if there aren't overlaps with purely local nodes, then the whole assumption of cause was incorrect and [here](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-9cfc4413865fc2a2f6942d9a215fc4e1R147) and [here](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-9cfc4413865fc2a2f6942d9a215fc4e1R160) in `datalib.py` were most likely the real culprits.

This is easy enough to test, as soon as I finish writing this commit.

Might I suggest for `storage.py`:

Something like this slotted into [line 114](https://github.com/deniszh/graphite-web/blob/53c6f615e8298e38a41ae5df0f27e432c5ed4449/webapp/graphite/storage.py#L114) instead:
```python
if settings.REMOTE_STORE_MERGE_RESULTS:
    for node in nodes_remaining:
        if not node.local:
            nodes_remaining.remove(best_node)
            minimal_node_set.add(node)
```

If however the *"there are no overlaps in coverage with purely local nodes"* holds true, then [your original method](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-57c0c4ffa18833a2e2ea55aab4309ffcR83) actually looks sound and probably didn't need to be changed.  It could even be one-upped to `if not remote_requests or settings.REMOTE_STORE_MERGE_RESULTS`.